Решение по 8 спринту

iliaalii opened the pull request:
- добавление в классы задач полей продолжительности и времени начала выполнения.
 - переписаны методы с учетом новых полей
 - в менеджер добавлен список приоритета задач
 - for заменен на stream api
 - добавлен абстрактный класс для тестов менеджеров
 - добавлены новые тесты покрывающие новые методы
 - в метод создания дефолтного файлового менеджера, сделана проверка на наличие файла, и его последующего создания при отсутствии

iliaalii committed:
Решение по 8 спринту
 - добавление в классы задач полей продолжительности и времени начала выполнения.
 - переписаны методы с учетом новых полей
 - в менеджер добавлен список приоритета задач
 - for заменен на stream api
 - добавлен абстрактный класс для тестов менеджеров
 - добавлены новые тесты покрывающие новые методы
 - в метод создания дефолтного файлового менеджера, сделана проверка на наличие файла, и его последующего создания при отсутствии

avfyodorov commented on src/controller/InMemoryTaskManager.java:
```diff
@@ -56,22 +63,29 @@ public void add(Epic epic) {
     @Override
     public void removeAllTask() {
+        tasks.values().forEach(prioritizedTasksList::remove);
         tasks.clear();
```
Ещё нужно не забыть удалить из истории.

iliaalii replied:
тоесть если задача более не доступна то и в истории она тоже не должна отображаться? 
просто не помню что бы в ТЗ было подобное,

iliaalii replied:
везде, где требовалось, добавил

avfyodorov replied:
Вот.. из ТЗ
.==
В приложении уже есть интерфейс для управления историей просмотров — HistoryManager. Осталось добавить метод void remove(int id) для удаления задачи из просмотра и реализовать его в классе InMemoryHistoryManager. **Добавьте вызов метода при удалении задач, чтобы они удалялись также из истории просмотров.**

avfyodorov commented on src/controller/InMemoryTaskManager.java:
```diff
@@ -56,22 +63,29 @@ public void add(Epic epic) {
+        tasks.values().forEach(prioritizedTasksList::remove);
         tasks.clear();
     }
 
```
public void removeAllEpic() {

Здесь также нужно удалить Эпики и  подзадачи из истории, А ещё подзадачи нужно удалить из отсортированного по времени списка prioritizedTasksList

avfyodorov commented on src/controller/InMemoryTaskManager.java:
```diff
@@ -83,11 +97,13 @@ public void removeAllEpic() {
     @Override
     public void removeAllSubtask() {
+        subtasks.values().forEach(prioritizedTasksList::remove);
         subtasks.clear();
```
И здесь нужно не забыть удалить подзадачи из менеджера истории.

avfyodorov commented on src/controller/InMemoryTaskManager.java:
```diff
@@ -122,91 +142,120 @@ public void update(Epic newEpic) {
     //Удаление по ID
     @Override
     public void removeTaskById(int taskId) {
+        prioritizedTasksList.remove(tasks.get(taskId));
```
и из истории тоже

avfyodorov commented on src/controller/InMemoryTaskManager.java:
```diff
@@ -122,91 +142,120 @@ public void update(Epic newEpic) {
+                        prioritizedTasksList.remove(subtasks.get(id));
+                        subtasks.remove(id);
+                        });
             epics.remove(epicId);
```
из истории тоже

avfyodorov commented on src/controller/InMemoryTaskManager.java:
```diff
@@ -122,91 +142,120 @@ public void update(Epic newEpic) {
+    }
+
+    @Override
+    public Set<Task> getPrioritizedTasks() {
```
Возвращать лучше список, а не множество... и копию. 
```java
	@Override
	public List<Task> getPrioritizedTasks() {
		return new ArrayList<>(prioritizedTasks);
	}
```

iliaalii replied:
изменил.
копию не стоит из за того, что нужно будет работать с задачами? или изменится сортировка?

avfyodorov reviewed (commented):
Добрый день, Илья! 

Хорошая работа, от меня всего пара уточнений.

iliaalii committed:
Исправления по 8 спринту
 - в методы с удалением, добавлено очистка из истории и списка приоритетов
 - изменен возвращаемый класс в методах поиска по id
 - исправил ошибку в обновлении подзадачи, ранее подзадачи дублировались в списке подзадач эпика
 - список приоритетов, теперь выводится списком, а не множеством
 - добавил тесты на проверку удалении истории и приоритетов

avfyodorov reviewed (commented):
Добрый день, Илья!

Замечаний нет.
Работа принята.
